test/test_hpystructseq.py

```diff
@@ -0,0 +1,132 @@
+"""
+NOTE: this tests are also meant to be run as PyPy "applevel" tests.
+
+This means that global imports will NOT be visible inside the test
+functions. In particular, you have to "import pytest" inside the test in order
+to be able to use e.g. pytest.raises (which on PyPy will be implemented by a
+"fake pytest module")
+"""
+from .support import HPyTest
+
+
+class TestHPyStructSequence(HPyTest):
+    def test_structseq(self):
+        import pytest
+        import sys
+        mod = self.make_module("""
+            static HPyStructSequence_Field structseq_fields[] = {
+                { "field0", "doc0" },
+                { "field1", NULL },
+                { NULL, NULL },
+                { NULL, NULL },
+            };
+            
+            static HPyStructSequence_Field structseq_empty_fields[] = {
+                { NULL, NULL },
+            };
+
+            static HPyStructSequence_Desc withfields_desc = {
+                .name = "mytest.WithFields",
+                .doc = "some doc",
+                .fields = structseq_fields
+            };
+
+            static HPyStructSequence_Desc nofields_desc = {
+                .name = "mytest.NoFields",
+                .fields = structseq_empty_fields
+            };
+            
+            HPyDef_METH(build, "build", HPyFunc_O)
+            static HPy build_impl(HPyContext *ctx, HPy self, HPy type)
+            {
+                HPyStructSequenceBuilder builder = HPyStructSequenceBuilder_New(ctx, type);
+                if (HPyStructSequenceBuilder_IsNull(builder))
+                    return HPy_NULL;
+                HPy tmp = HPyLong_FromLong(ctx, 1);
+                HPyStructSequenceBuilder_Set(ctx, builder, 0, tmp);
+                HPy_Close(ctx, tmp);
+                HPyStructSequenceBuilder_Set_i(ctx, builder, 1, 2);
+                HPyStructSequenceBuilder_Set_i(ctx, builder, 2, 3);
+                return HPyStructSequenceBuilder_Build(ctx, builder, type);
```

Comment:
What happens if a different `type` is used here? Shouldn't the `type` be redundant since it was used in `HPyStructSequenceBuilder_New`?

Reply:
> What happens if a different type is used here?

If the types then have a different number of elements, construction would not work. Depending on the ABI mode, it would maybe just crash or raise an error. Hence, behavior is undefined.

I agree, it shouldn't be necessary to pass the type to `HPyStructSequenceBuilder_Build`. And yes, behavior should be defined and consistent. Sorry, I was already a bit tired of merging behavior of `namedtuple` and `structseq`.